align/schema/constraint.py

```diff
@@ -1,155 +1,257 @@
+                        checker.And(b2.lly >= b1.lly, b2.ury <= b1.ury)
+                    )
+                )
 
-    def __init__(self, validation=None):
-        super().__init__(__root__=[])
-        self._commits = collections.OrderedDict()
-        if z3 is not None:
-            self._solver = z3.Solver()
-            if validation is not None:
-                self._validation = validation
-        else:
-            assert not validation, "Cannot validate without z3"
+# You may chain constraints together for more complex constraints by
+#     1) Assigning default values to certain attributes
+#     2) Using custom validators to modify attribute values
+# Note: Compositional check() is automatically constructed if
+#     every check() in mro starts with `constraints = super().check()`.
+#     (mro is Order, Align, ConstraintBase in this example)
+# Note: If you need to specialize check(), you do have the option 
+#     to create a custom `check()` in this class. It shouldn't be
+#     needed unless you are adding new semantics
+class AlignInOrder(Order, Align):
+    '''
+    Align `instances` in order, along the `direction`, on the `line`
+
+    > `direction == 'horizontal'` => left_to_right
+
+    > `direction == 'vertical'`   => bottom_to_top
+    '''
+    instances : List[str]
+    direction: Optional[Literal['horizontal', 'vertical']]
+    line : Optional[Literal[
+        'top', 'bottom',
+        'left', 'right',
+        'center'
+    ]]
+    abut: Optional[bool] = False
+
+    @types.root_validator(allow_reuse=True)
+    def _cast_constraints_to_base_types(cls, values):
+        # Process unambiguous line values
+        if values['line'] in ['bottom', 'top']:
+            if values['direction'] is None:
+                values['direction'] = 'horizontal'
+            else:
+                assert values['direction'] == 'horizontal', \
+                    f'direction is horizontal if line is bottom or top'
+            values['line'] = f"h_{values['line']}"
+        elif values['line'] in ['left', 'right']:
+            if values['direction'] is None:
+                values['direction'] = 'vertical'
+            else:
+                assert values['direction'] == 'vertical', \
+                    f'direction is vertical if line is left or right'
+            values['line'] = f"v_{values['line']}"
+        # Center needs both line & direction
+        elif values['line'] == 'center':
+            assert values['direction'], \
+                'direction must be specified if line == center'
+            values['line'] = f"{values['direction'][0]}_{values['line']}"
+        # Map horizontal, vertical direction to left_to_right & bottom_to_top
+        if values['direction'] == 'horizontal':
+            values['direction'] = 'left_to_right'
+        elif values['direction'] == 'vertical':
+            values['direction'] = 'bottom_to_top'
+        return values
+
+ConstraintType=Union[Order, Align, AlignInOrder]
```

Comment:
1. Where is `Generate`?
2. End user can employ any of these 3 constraints 👍 , right?

Reply:
I will work on `generate` in a separate PR once I start reworking the schematic2layout flow. The validator / checker for generate should check that the generator exists.

End user can use all three right now yes.

Reply:
👍